File: Week 8/cpp/1.binary_search.cpp
/*
 * WEEK 8 - C++ DSA
 * Topic: Binary Search
 * File: 1.binary_search.cpp
 *
 * CONCEPT:
 *     Find a value in a sorted vector in O(log n) by halving the search
 *     interval each step.
 *
 * KEY POINTS:
 *     - PREREQUISITE: vector must be sorted.
 *     - Compute mid as low + (high - low) / 2 to avoid overflow.
 *     - Variants: first/last occurrence, count, rotated-array search.
 *
 * ALGORITHM / APPROACH:
 *     Standard iterative:
 *         low = 0; high = n - 1
 *         while (low <= high) {
 *             mid = low + (high - low) / 2
 *             if arr[mid] == target return mid
 *             else if arr[mid] < target low = mid + 1
 *             else high = mid - 1
 *         }
 *         return -1
 *
 * C++-SPECIFIC NOTES:
 *     - <algorithm> provides std::lower_bound / std::upper_bound /
 *       std::binary_search; we implement manually for instruction.
 *     - Be careful with size_t vs int when high may go to -1.
 *
 * DRY RUN:
 *     [-5,-2,0,1,3,5,7,9,11], target=5:
 *         low=0 high=8 mid=4 arr[4]=3 < 5 -> low=5
 *         low=5 high=8 mid=6 arr[6]=7 > 5 -> high=5
 *         low=5 high=5 mid=5 arr[5]=5 -> return 5
 *
 * COMPLEXITY:
 *     Time:  O(log n)
 *     Space: O(1) iterative, O(log n) recursive
 */

#include <iostream>
#include <vector>
#include <algorithm>

int binarySearch(const std::vector<int>& arr, int target) {
    int low = 0, high = static_cast<int>(arr.size()) - 1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (arr[mid] == target) return mid;
        else if (arr[mid] < target) low = mid + 1;
        else high = mid - 1;
    }
    return -1;
}

int binarySearchRec(const std::vector<int>& arr, int target, int low, int high) {
    if (low > high) return -1;
    int mid = low + (high - low) / 2;
    if (arr[mid] == target) return mid;
    else if (arr[mid] < target) return binarySearchRec(arr, target, mid + 1, high);
    else return binarySearchRec(arr, target, low, mid - 1);
}

int firstOccurrence(const std::vector<int>& arr, int target) {
    int low = 0, high = static_cast<int>(arr.size()) - 1, result = -1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (arr[mid] == target) { result = mid; high = mid - 1; }
        else if (arr[mid] < target) low = mid + 1;
        else high = mid - 1;
    }
    return result;
}

int lastOccurrence(const std::vector<int>& arr, int target) {
    int low = 0, high = static_cast<int>(arr.size()) - 1, result = -1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (arr[mid] == target) { result = mid; low = mid + 1; }
        else if (arr[mid] < target) low = mid + 1;
        else high = mid - 1;
    }
    return result;
}

int countOccurrences(const std::vector<int>& arr, int target) {
    int first = firstOccurrence(arr, target);
    if (first == -1) return 0;
    return lastOccurrence(arr, target) - first + 1;
}

int searchRotated(const std::vector<int>& arr, int target) {
    int low = 0, high = static_cast<int>(arr.size()) - 1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (arr[mid] == target) return mid;
        if (arr[low] <= arr[mid]) {                      // left half sorted
            if (arr[low] <= target && target < arr[mid]) high = mid - 1;
            else                                          low  = mid + 1;
        } else {                                          // right half sorted
            if (arr[mid] < target && target <= arr[high]) low  = mid + 1;
            else                                          high = mid - 1;
        }
    }
    return -1;
}

int main() {
    std::vector<int> sorted = {-5, -2, 0, 1, 3, 5, 7, 9, 11};
    std::cout << "Array: [-5,-2,0,1,3,5,7,9,11]\n";
    std::cout << "binarySearch(5) = " << binarySearch(sorted, 5) << "\n";
    std::cout << "binarySearch(0) = " << binarySearch(sorted, 0) << "\n";
    std::cout << "binarySearch(4) = " << binarySearch(sorted, 4) << "\n";
    std::cout << "binarySearchRec(7) = "
              << binarySearchRec(sorted, 7, 0, static_cast<int>(sorted.size()) - 1) << "\n";

    std::vector<int> withDups = {1, 2, 2, 2, 3, 4, 4, 5};
    std::cout << "\nArray with duplicates: [1,2,2,2,3,4,4,5]\n";
    std::cout << "firstOccurrence(2) = " << firstOccurrence(withDups, 2) << "\n";
    std::cout << "lastOccurrence(2)  = " << lastOccurrence(withDups, 2) << "\n";
    std::cout << "countOccurrences(2) = " << countOccurrences(withDups, 2) << "\n";
    std::cout << "countOccurrences(4) = " << countOccurrences(withDups, 4) << "\n";
    std::cout << "countOccurrences(6) = " << countOccurrences(withDups, 6) << "\n";

    std::vector<int> rotated = {4, 5, 6, 7, 0, 1, 2};
    std::cout << "\nRotated array: [4,5,6,7,0,1,2]\n";
    std::cout << "searchRotated(0) = " << searchRotated(rotated, 0) << "\n";
    std::cout << "searchRotated(6) = " << searchRotated(rotated, 6) << "\n";
    std::cout << "searchRotated(3) = " << searchRotated(rotated, 3) << "\n";

    // STL alternative
    auto it = std::lower_bound(sorted.begin(), sorted.end(), 5);
    std::cout << "\nlower_bound(5) idx = " << (it - sorted.begin()) << "\n";

    return 0;
}

/*
 * NOTES — C++ vs Java:
 *     - <algorithm> already has std::binary_search, std::lower_bound,
 *       std::upper_bound — the standard library equivalents.
 *     - Avoid using size_t arithmetic that can underflow; cast to int for
 *       indices that may decrease past zero.
 */

File: Week 8/rust/1.binary_search.rs
/*
 * WEEK 8 - RUST DSA
 * Topic: Binary Search
 * File: 1.binary_search.rs
 *
 * CONCEPT:
 *     Locate a value in a sorted slice in O(log n) by halving the
 *     interval.
 *
 * KEY POINTS:
 *     - Slice must be sorted.
 *     - Use i64/isize for half-open arithmetic when high may go below 0.
 *     - Variants: first/last occurrence, count, search in rotated array.
 *
 * ALGORITHM / APPROACH:
 *     Standard iterative as in Java/C++/Python files.
 *
 * RUST-SPECIFIC NOTES:
 *     - The standard library offers slice::binary_search returning
 *       Result<usize, usize>: Ok(index) or Err(insert_position).
 *     - Convert to Option<usize> or -1 sentinel for parity with Java.
 *
 * DRY RUN:
 *     [-5,-2,0,1,3,5,7,9,11], target=5 -> index 5.
 *     Rotated [4,5,6,7,0,1,2], target=0 -> index 4.
 *
 * COMPLEXITY:
 *     Time:  O(log n)
 *     Space: O(1) iterative, O(log n) recursive
 */

fn binary_search(arr: &[i32], target: i32) -> i64 {
    let mut low: i64 = 0;
    let mut high: i64 = arr.len() as i64 - 1;
    while low <= high {
        let mid = low + (high - low) / 2;
        let v = arr[mid as usize];
        if v == target {
            return mid;
        } else if v < target {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    -1
}

fn binary_search_rec(arr: &[i32], target: i32, low: i64, high: i64) -> i64 {
    if low > high {
        return -1;
    }
    let mid = low + (high - low) / 2;
    let v = arr[mid as usize];
    if v == target {
        mid
    } else if v < target {
        binary_search_rec(arr, target, mid + 1, high)
    } else {
        binary_search_rec(arr, target, low, mid - 1)
    }
}

fn first_occurrence(arr: &[i32], target: i32) -> i64 {
    let mut low: i64 = 0;
    let mut high: i64 = arr.len() as i64 - 1;
    let mut result: i64 = -1;
    while low <= high {
        let mid = low + (high - low) / 2;
        let v = arr[mid as usize];
        if v == target {
            result = mid;
            high = mid - 1;
        } else if v < target {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    result
}

fn last_occurrence(arr: &[i32], target: i32) -> i64 {
    let mut low: i64 = 0;
    let mut high: i64 = arr.len() as i64 - 1;
    let mut result: i64 = -1;
    while low <= high {
        let mid = low + (high - low) / 2;
        let v = arr[mid as usize];
        if v == target {
            result = mid;
            low = mid + 1;
        } else if v < target {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    result
}

fn count_occurrences(arr: &[i32], target: i32) -> i64 {
    let first = first_occurrence(arr, target);
    if first == -1 {
        return 0;
    }
    last_occurrence(arr, target) - first + 1
}

fn search_rotated(arr: &[i32], target: i32) -> i64 {
    let mut low: i64 = 0;
    let mut high: i64 = arr.len() as i64 - 1;
    while low <= high {
        let mid = low + (high - low) / 2;
        let v = arr[mid as usize];
        if v == target {
            return mid;
        }
        if arr[low as usize] <= v {
            // left half sorted
            if arr[low as usize] <= target && target < v {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        } else {
            // right half sorted
            if v < target && target <= arr[high as usize] {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
    }
    -1
}

fn main() {
    let sorted = vec![-5, -2, 0, 1, 3, 5, 7, 9, 11];
    println!("Array: {:?}", sorted);
    println!("binary_search(5) = {}", binary_search(&sorted, 5));
    println!("binary_search(0) = {}", binary_search(&sorted, 0));
    println!("binary_search(4) = {}", binary_search(&sorted, 4));
    println!(
        "binary_search_rec(7) = {}",
        binary_search_rec(&sorted, 7, 0, sorted.len() as i64 - 1)
    );

    let with_dups = vec![1, 2, 2, 2, 3, 4, 4, 5];
    println!("\nArray with duplicates: {:?}", with_dups);
    println!("first_occurrence(2) = {}", first_occurrence(&with_dups, 2));
    println!("last_occurrence(2)  = {}", last_occurrence(&with_dups, 2));
    println!("count_occurrences(2) = {}", count_occurrences(&with_dups, 2));
    println!("count_occurrences(4) = {}", count_occurrences(&with_dups, 4));
    println!("count_occurrences(6) = {}", count_occurrences(&with_dups, 6));

    let rotated = vec![4, 5, 6, 7, 0, 1, 2];
    println!("\nRotated array: {:?}", rotated);
    println!("search_rotated(0) = {}", search_rotated(&rotated, 0));
    println!("search_rotated(6) = {}", search_rotated(&rotated, 6));
    println!("search_rotated(3) = {}", search_rotated(&rotated, 3));

    // Standard library
    let std = sorted.binary_search(&5);
    println!("\nslice::binary_search(5) = {:?}", std);
}

/*
 * NOTES — Rust vs Java:
 *     - The slice API offers binary_search natively returning Result.
 *     - usize underflows at 0; we use i64 indices to safely handle "high = -1".
 *     - Pattern matching makes Result/Option-based standard search ergonomic.
 */
